### PR TITLE
Storing the `NODE_FEATURES` in the annotation graph once for speed-up

### LINKED ISSUE #276: [FEATURE REQUEST] Append node features to the graph once and for all
https://github.com/alan-turing-institute/grace/blob/4feae20cadcdea63b4e819b65de30ded6b7770a2/grace/simulator/README.md?plain=1#L32-L38

### PR DESCRIPTION
**Pull request by KristinaUlicna**
## PR contribution summary

**Why is this PR useful / good for?** *Please describe the problem(s) you're trying to address.*
* Script to append all `.grace` folders in the given path with `nodes.parquet` file which holds the `NODE_FEATURES`
* `GraphAttrs.NODE_FEATURES` are now in the graph so that the feature extraction doesn't have to be repeated at every run
* If you don't want to re-run the feature extraction, supply `None` (default) to the `config.extractor_fn`

### List of proposed changes / linked issues & discussions

- [x] ✅ Resolves #276 

### What should a reviewer concentrate their feedback on?

+ 🏃 Scripts to run

https://github.com/alan-turing-institute/grace/blob/4feae20cadcdea63b4e819b65de30ded6b7770a2/grace/simulator/README.md?plain=1#L32-L38

+ 📝 Everything looks OK?

#### What type of PR is this? (check all applicable)

+ 🪄 Feature
+ #️⃣ Documentation / code annotation
+ 🧑‍💻 Code refactor / style

#### Added tests?

+ 👍 yes
+ 🙅 no, because they aren't needed
+ 🙋 no, because I need some help

---

## PR review summary

Describe what this PR does & **how you reviewed** the individual items, where needed:

* <TEXT>

*Some helper checks to tick off:*

* Focus on image annotation
* Focus on model training
* Could any optimization be applied?
* Is there any redundant code?
* Are there any spelling errors?

**In conclusion**, after my review, I'd like to:

* 🙋 ask some clarifying questions
* 🙅 suggest some specific changes

### CODE REVIEW

File: grace/simulator/store_features.py
Comment (crangelsmith) [inline grace/simulator/store_features.py:1]: I wonder why do we prefer to do this in a separate script? Woudnt be easier that on the first run (if no graph features is provided as an input) the features are extracted and saved for a following run?

File: grace/run.py [hunk 1]
```diff
@@ -56,28 +56,35 @@ def run_grace(config_file: Union[str, os.PathLike]) -> None:
...
     # Condition the augmentations to train mode only:
     def transform(
         image: torch.Tensor, graph: dict, *, in_train_mode: bool = True
     ) -> Callable:
-        # Ensure augmentations are only run on train data:
-        if in_train_mode:
-            image_aug, graph_aug = img_graph_augs(image, graph)
-            return feature_extractor(image_aug, graph_aug)
+        # Prepare the feature extractor:
+        if config.extractor_fn is not None:
```
Comment (crangelsmith) [inline grace/run.py:71]: So the logic here is that the features are part of the loaded graph and is we don't want to compute new features we don't provided `extractor_fn` in the config file?

File: grace/simulator/store_features.py
Comment (KristinaUlicna) [reply grace/simulator/store_features.py:1]: I think this choice limits our ability to do patch / graph augmentations on the go. I think the entire approach of how I create the synthetic dataset need to be rebuild to allow some flexibility (PR incoming soon). This is a temporary hack that works for now, but I'll need to think about how not to jeopardise the option to include augmentations in the process.

File: grace/run.py (hunk 1 above)
Comment (KristinaUlicna) [reply grace/run.py:71]: Exactly - just temporarily though, because none of my experiments include patch / graph augmentations (yet).

**Comment by crangelsmith** [comment]
Ok, so i think is ok to merge as it is something that will allow you to quickly iterate on the model improvement. However, should have an eye on how we want the pipeline to look for an external user and what to keep inside or outside of it.

**Pull request closed without merging**